app/actions/views/file_upload.js

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import FormData from 'form-data';
+import {Platform} from 'react-native';
+import {uploadFile} from 'mattermost-redux/actions/files';
+import {lookupMimeType} from 'mattermost-redux/utils/file_utils';
+
+import {generateId} from 'app/utils/file';
+import {ViewTypes} from 'app/constants';
+
+export function handleUploadFiles(files, rootId, requestId) {
+    return async (dispatch, getState) => {
+        const state = getState();
+
+        const teamId = state.entities.teams.currentTeamId;
+        const channelId = state.entities.channels.currentChannelId;
+        const formData = new FormData();
+
+        files.forEach((file) => {
+            const name = file.path.split('/').pop();
+            const mimeType = lookupMimeType(name);
+            const clientId = generateId();
+
+            const fileData = {
+                uri: file.path,
+                name,
+                type: mimeType
+            };
+
+            formData.append('files', fileData);
+            formData.append('channel_id', channelId);
+            formData.append('client_ids', clientId);
+        });
+
+        let formBoundary;
+        if (Platform.os === 'ios') {
+            formBoundary = '--mobile.client.file.upload';
```

Comment:
Taking some hints from this thread: https://github.com/facebook/react-native/issues/5308

Removing the boundary on Android was the only way I could get the upload to work. I think it's an internal issue with Fetch on Android. I'm still doing some investigation on it, but this was the way I could get it to work for now.